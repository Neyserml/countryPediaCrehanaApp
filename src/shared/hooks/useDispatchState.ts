import { useContext, Context, DispatchWithoutAction } from 'react';

export type DispatchState<T = any, D = DispatchWithoutAction> = {
  dispatch: D;
  state: T;
};

export default function useDispatchState<T extends DispatchState<T['state'], T['dispatch']>>(
  context: Context<T>,
): [T['state'], T['dispatch']] {
  const { state, dispatch } = useContext(context);
  return [state, dispatch];
}
