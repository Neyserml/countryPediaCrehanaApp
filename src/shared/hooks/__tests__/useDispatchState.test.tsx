import { render } from "@testing-library/react";
import React, { useReducer } from "react";
import useDispatchState, { DispatchState } from "../useDispatchState";

interface MyDemoState {
  title: string;
}

type DemoContextType = DispatchState<Partial<MyDemoState>>;

const DemoContext = React.createContext({} as DemoContextType);

const DemoContextProvider: React.FC = ({ children }) => {
  const [state, dispatch] = useReducer(() => ({}), {
    title: "crehana",
  });
  return (
    <DemoContext.Provider value={{ dispatch, state }}>
      {children}
    </DemoContext.Provider>
  );
};

describe("useDispatchState", () => {
  it("should provide dispatch and state values", () => {
    const DemoComponent: React.FC = () => {
      const [state, dispatch] = useDispatchState(DemoContext);

      expect(state.title).toBe("crehana");
      expect(dispatch).toBeInstanceOf(Function);

      return <strong>{state.title}</strong>;
    };

    const component = render(
      <DemoContextProvider>
        <DemoComponent />
      </DemoContextProvider>
    );

    expect(component.container).toMatchSnapshot();
  });
});
