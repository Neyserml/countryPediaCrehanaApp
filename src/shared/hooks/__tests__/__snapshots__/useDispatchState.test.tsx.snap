// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useDispatchState should provide dispatch and state values 1`] = `
<div>
  <strong>
    crehana
  </strong>
</div>
`;
